StateDB review follow-ups

joamaki opened the pull request:
Follow-ups to the original statedb v2.0 PR (https://github.com/cilium/cilium/pull/27160).

This addresses the original review comments, expands tests to improve coverage (81.8% => 86.1%) and
fixes few metrics related issues and adds tests for gauge metrics.

joamaki committed:
statedb: Extend DeleteTracker tests for full coverage

Expand the test case to fully cover the implementation.
Use a second DeleteTracker to validate that the second tracker
holds back garbage collection.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Add tests for Revision()

Add tests to cover Revision() for both ReadTxn and WriteTxn
cases.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Remove txn from Mark()

This wasn't required and was a leftover from when the Mark did not
take a revision, but rather derived the revision from the transaction.
This earlier approach didn't work as it wouldn't have allowed partial
consuming of the deleted objects.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Use a short graveyard GC interval in tests

To avoid delays in tests, set the graveyard GC interval to
a low value.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Add errors.go to collect the errors in one place

Allows use of "errors.Is" to pattern match on the returned errors.
Add test for Commit() and Abort() to validate the use of some of the new errors.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb/index: Add comment on BigEndian indexing and fix test

Addresses review comments from original PR to expand
on why we want BigEndian byte order for integers and fix
small issue in keyset test.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Clean up use of getTable and address review comments

getTable() method was poorly defined, it returned the currently
mutating table if the table was targeted in write transaction,
or the immutable version if not. It was mostly used with DeleteTracker
and not elsewhere.

To make it more obvious, remove the method, rename "tables" to "modifiedTables"
and look up the table from there when "deleteTrackers" needs to be modified.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Cleanup tests by dedup graveard empty check and use t.Cleanup

The require.Eventually for graveyard emptyness was duplicated with different
timeouts. Benchmark had way too low timeout leading to failure. Dedup and
use reasonably high timeout.

Instead of calling a callback and using defer to cleanup, use t.Cleanup
instead to avoid nesting in every test case.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Add documentation and comments for Abort/Commit/DeleteAll

From PR review, expand the comments in Abort and Commit to explain
properties/assumptions and behaviour.

Add missing doc for DeleteAll.

Add why nolint is used on Index[] methods.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/statedb/types.go:
```diff
@@ -56,6 +56,19 @@ type Table[Obj any] interface {
+	// If a suitable WriteTxn is provided, then that is used to add the tracker.
+	// In that case the tracker will only become active when the transaction is
+	// committed.
+	DeleteTracker(txn ReadTxn, trackerName string) (*DeleteTracker[Obj], error)
```
Or we just keep it as `WriteTxn` and then `DeleteTracker` is the only usable method that someone without `RWTable` can use to "modify the table", e.g. readers who have `Table[Obj]` can do `WriteTxn(table)` and call `DeleteTracker` with it. Would get rid of this deadlock potential.

I think it's a reasonable trade-off to still expose `WriteTxn` against even read-only tables for everyone as it allows this DeleteTracker() stuff and it allows building smart writers on top of `RWTable[Obj]` while still allowing cross-table transactionality. The downside I see is that then arbitrary code can "lock" the table even when they can't modify it otherwise. Is this practical engineering or am I being reckless?

joamaki committed:
statedb: Improve metrics

- Fixes to bugs from the original review in https://github.com/cilium/cilium/pull/27160.
- Remove the pending* maps from txn and update the object counts based on Tree.Len()
- Add checks into tests for gauge metrics
- Remove _total suffix from metrics

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>
Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on pkg/statedb/db_test.go:
```diff
@@ -91,517 +89,504 @@ func testWithDB(t testing.TB, withTags bool, test func(db *DB, table Table[testO
+	return db, table
 }
 
 func TestDB_LowerBound_ByRevision(t *testing.T) {
```
Could some of these be https://pkg.go.dev/testing#T.Parallel? 
(Was reminded about this due to the Cleanup thing, https://brandur.org/fragments/go-prefer-t-cleanup-with-parallel-subtests)

joamaki replied:
Yes they can. I'll play with it.

joamaki replied:
Done. `0.346s` to `0.210s`.

bimmlerd reviewed (approved):
LGTM

joamaki committed:
statedb: Mark applicable tests parallel

Most tests are lean and can be run in parallel to each other
to speed up testing.

Before:
ok  	github.com/cilium/cilium/pkg/statedb	0.346s

After:
ok  	github.com/cilium/cilium/pkg/statedb	0.210s



Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Move benchmarks to benchmarks_test.go

The db_test.go is starting to get a bit too big.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink reviewed (approved):
LGTM
